subprojects/rzwinkd/winkd.c

```diff
@@ -351,231 +361,240 @@ int winkd_walk_vadtree(WindCtx *ctx, ut64 address, ut64 parent) {
+		if (proc) {
+			rz_list_append(ret, proc);
+		}
 		ptr = next;
-	} while (ptr != base);
-
-	ctx->plist_cache = ret;
-
+	} while (ptr != ctx->PsActiveProcessHead);
 	return ret;
 }
 
-int winkd_write_at_uva(WindCtx *ctx, const uint8_t *buf, ut64 offset, int count) {
-	ut64 pa;
-	ut32 totwritten = 0;
-	while (totwritten < count) {
-		if (!winkd_va_to_pa(ctx, offset, &pa)) {
-			return 0;
+int winkd_op_at_uva(WindCtx *ctx, ut64 address, uint8_t *buf, int count, bool write) {
+	ut32 total = 0;
+	ut32 offset = 0;
+	const ut64 end = address + count;
+	while (address < end) {
+		ut64 pa;
+		const ut32 restOfPage = 0x1000 - (address & 0xfff);
+		if (!winkd_va_to_pa(ctx, ctx->target.dir_base_table, address, &pa)) {
+			address += restOfPage;
+			offset += restOfPage;
+			continue;
 		}
-		ut32 restOfPage = 0x1000 - (offset & 0xfff);
-		int written = winkd_write_at_phys(ctx, buf + totwritten, pa, RZ_MIN(count - totwritten, restOfPage));
-		if (!written) {
-			break;
+		int result;
+		if (write) {
+			result = ctx->write_at_physical(ctx->user, pa, buf + offset, RZ_MIN(count - offset, restOfPage));
+		} else {
+			result = ctx->read_at_physical(ctx->user, pa, buf + offset, RZ_MIN(count - offset, restOfPage));
 		}
-		offset += written;
-		totwritten += written;
+		address += restOfPage;
+		offset += restOfPage;
+		total += result;
 	}
-	return totwritten;
+	return total;
 }
 
-int winkd_read_at_uva(WindCtx *ctx, uint8_t *buf, ut64 offset, int count) {
-	ut64 pa;
-	ut32 totread = 0;
-	while (totread < count) {
-		if (!winkd_va_to_pa(ctx, offset, &pa)) {
-			return 0;
-		}
-		ut32 restOfPage = 0x1000 - (offset & 0xfff);
-		int read = winkd_read_at_phys(ctx, buf + totread, pa, RZ_MIN(count - totread, restOfPage));
-		if (!read) {
-			break;
-		}
-		offset += read;
-		totread += read;
-	}
-	return totread;
+int winkd_read_at_uva(WindCtx *ctx, ut64 address, uint8_t *buf, int count) {
+	return winkd_op_at_uva(ctx, address, buf, count, false);
+}
+
+int winkd_write_at_uva(WindCtx *ctx, ut64 address, const uint8_t *buf, int count) {
+	return winkd_op_at_uva(ctx, address, (uint8_t *)buf, count, true);
 }
 
 RzList *winkd_list_modules(WindCtx *ctx) {
-	RzList *ret;
 	ut64 ptr, base;
 
-	if (!ctx || !ctx->desc || !ctx->syncd) {
-		return NULL;
-	}
-
-	if (!ctx->target) {
+	if (!ctx->target.uniqueid) {
 		eprintf("No target process\n");
 		return NULL;
 	}
+	int list_entry_off = 0;
+	const bool is_target_kernel = ctx->target.uniqueid == 4;
+	if (is_target_kernel) {
+		if (!ctx->PsLoadedModuleList) {
+			eprintf("No PsLoadedModuleList\n");
+			return NULL;
+		}
+		ptr = ctx->PsLoadedModuleList;
+		base = ptr;
+		if (!winkd_read_at_uva(ctx, ptr, (uint8_t *)&ptr, 4 << ctx->is_64bit)) {
+			eprintf("PsLoadedModuleList not present\n");
+		}
+		if (ptr == base) {
+			return NULL;
+		}
+	} else {
+		ut64 ldroff = ctx->is_64bit ? 0x18 : 0xC;
+		list_entry_off = (4 << ctx->is_64bit) * 2;
+		if (!ctx->target.peb) {
+			eprintf("No PEB\n");
```

Comment:
`RZ_LOG_ERROR()`

Reply:
Maybe also worth adding a comment that this piece of code does, since there are magical offsets here